pkg/steps/template.go

```diff
@@ -422,23 +422,65 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 }
 
 func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
-	timeout := 300
-	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
-		}
-		if err != nil {
-			return fmt.Errorf("could not retrieve deleting pod: %v", err)
-		}
-		if pod.UID != uid {
-			return nil
+	hasTeardownContainer := false
+	timeout := 5 * time.Minute
+	teardownTimeout := 10 * time.Minute
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	switch {
+	case errors.IsNotFound(err), pod.UID != uid:
+		// Pod was removed or mismatching UID
+		return nil
+	case err != nil:
+		return fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+
+	// Attempts to wait for teardown to complete
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Name == "teardown" {
+			hasTeardownContainer = true
+			timeout = timeout + teardownTimeout
+			break
 		}
-		log.Printf("Waiting for pod %s to be deleted ... (%ds/%d)", name, i, timeout)
-		time.Sleep(2 * time.Second)
 	}
 
-	return fmt.Errorf("waited for pod %s deletion for %ds, was not deleted", name, timeout)
+	watcher, err := podClient.Watch(meta.ListOptions{
+		FieldSelector: fields.Set{"metadata.name": name}.AsSelector().String(),
+		Watch:         true,
+	})
+	if err != nil {
+		return fmt.Errorf("could not create watcher for pod: %v", err)
+	}
+	defer watcher.Stop()
+
+	log.Printf("Waiting for pod %s to be deleted in %d seconds", name, timeout)
+	_, err = watch.Until(timeout, watcher, func(event watch.Event) (done bool, err error) {
+		switch event.Type {
+		case watch.Deleted:
+			// pod was deleted
+			return true, nil
+		case watch.Added, watch.Modified:
+			if hasTeardownContainer {
```

Comment:
I don't understand this logic. If we have a `teardown` container, we will exit out early every time. If we _don't_ have a `teardown` container, we set this boolean to `false`. The comment says that will avoid re-checking, but in reality that means we check every time. Then, if the `teardown` container is terminated, you signal you are done, so the watch ends. I think we just need a dead-simple watch, or two watches. If you want to have _one_ watch with a variable timeout, just wait for the deletion. If you want to wait for the `teardown` container completion and the pod deletion separately, you will want separate watches. 

In general, if the issue was a too-short timeout that cut the `teardown` container short, why not just make this watch go on for an hour? In what cases do we _not_ want to wait for the Pod to really be gone?

Also, if you look at the implementations in the build utils, we want a list then a watch with retires to handle transient errors.

Reply:
>If we have a teardown container, we will exit out early every time

Fixed by introducing `teardownFinished` var

>If you want to wait for the teardown container completion and the pod deletion separately, you will want separate watches.

That was my initial idea (see f758b7df804c1f9bdf7a796ad9a762acc284c784), however there is short time (between teardown watch and pod to be deleted watch) where pod may be destroyed and could have been replaced with a new pod. Two watches don't seem reliable to me

>why not just make this watch go on for an hour? In what cases do we not want to wait for the Pod to really be gone?

That would hide potential issues in teardown

>we want a list then a watch with retires to handle transient errors.

Using `event, ok := <-watcher.ResultChan()`? It doesn't seem to have some kind of timeout